### PR TITLE
Typecheck record access

### PR DESCRIPTION
**Pull request by SebastianMestre**
I don't think this is fine. I get the feeling I missed a check in func_unify, but I don't know what it is.

### CODE REVIEW

**Comment by SebastianMestre** [comment]
I think it's good now.

It's not pretty, though

File: src/typesystem.hpp [hunk 1]
```diff
@@ -17,9 +17,9 @@ enum class TypeFunctionTag { Var, Builtin, Sum, Product, Record };
 // For the purposes of the type system, we only care about the amount
 // of argument it takes.
 
-// If it's a var, equals points to another TypeFunctionData.
+// If it's a var, equals points to another TypeFunctionHeader.
 // Else it points to a TypeFunction.
-struct TypeFunctionData {
+struct TypeFunctionHeader {
```
Comment (lushoBarlett) [inline src/typesystem.hpp:22]: I'm all for this name. I really like it.

File: src/typesystem.cpp [hunk 2]
```diff
@@ -208,17 +208,69 @@ void TypeSystemCore::func_unify(TypeFunctionId a, TypeFunctionId b) {
...
-		b_header.equals = a;
-	else
-		assert(0 and "unifying two different known type functions");
-}
+	// ensure a is a var if at least one of them is a var
+	if (type_function_header[b].type == TypeFunctionTag::Var)
+		std::swap(a, b);
+
+	if (type_function_header[a].type == TypeFunctionTag::Var) {
+		type_function_header[a].equals = b;
```
Comment (lushoBarlett) [inline src/typesystem.cpp:220]: I like this style, I'll be sure to keep `std::swap` in mind for the future.

File: src/typesystem.cpp [hunk 3]
```diff
@@ -208,17 +208,69 @@ void TypeSystemCore::func_unify(TypeFunctionId a, TypeFunctionId b) {
...
+			// this way, if more fields get added to a or b, both get updated
+			//
+			// TODO: think more thoroughly about this...
+			// I get the feeling that we shouldn't really do this if b is a
+			// polymorphic record. why? Because b will have type variables that
+			// are bound to a forall kinda thing, and thus should NOT be unified.
+			// instead, we should do something similar to inst_fresh, but for
+			// type funcs.
+			type_function_header[a].type = TypeFunctionTag::Var;
+			type_function_header[a].equals = b;
```
Comment (lushoBarlett) [inline src/typesystem.cpp:270]: What do you mean?

Comment (SebastianMestre) [reply src/typesystem.cpp:270]: if we have a dummy-record in the type of a top-level declaration, users of that declaration can unify stuff with it and modify it.

This is not ideal if we want to have polymorphism on records.

File: src/typecheck.cpp [hunk 4]
```diff
@@ -226,14 +226,18 @@ void typecheck(TypedAST::TernaryExpression* ast, Frontend::CompileTimeEnvironmen
 }
 
 void typecheck(TypedAST::RecordAccessExpression* ast, Frontend::CompileTimeEnvironment& env) {
+	typecheck(ast->m_record.get(), env);
```
Comment (lushoBarlett) [inline src/typecheck.cpp:231]: Is this function defined somewhere?

**Review by lushoBarlett — APPROVED** [review]
I think this is correct.

I'm not sure if a test is possible for this yet. If it is, please consider adding it.

File: src/typecheck.cpp (hunk 4 above)
Comment (SebastianMestre) [reply src/typecheck.cpp:231]: `m_record.get()` is a `TypedAST*`.

`typecheck(TypedAST*, CompileTimeEnvironment&)` is the very last definition in this file.

File: src/typesystem.cpp (hunk 3 above)
Comment (SebastianMestre) [reply src/typesystem.cpp:270]: Essentially, dummy records don't get generalized, so they can be mutated after we are done typechecking the declaration they came from.

**Comment by SebastianMestre** [comment]
> I think this is correct.
> 
> I'm not sure if a test is possible for this yet. If it is, please consider adding it.

Uhh... kinda? I'll add a test, you let me know if it actually helps at all.

**Pull request merged**